### PR TITLE
Network: abstract dbus and supervisor - ipv6/wifi/vlan

### PR DESCRIPTION
**Pull request by pvizeli**
## Proposed change

- Detect all network cards, also if they are not active or used (solve issue with unsupported without network)
- Support IPv4/IPv6
- Support Wifi + Scan
- Support VLAN (Let's allow people do it right, also if they have no plan about)
- Support enable/disable interfaces

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to the supervisor)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes # https://github.com/home-assistant/operating-system/issues/933
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast supervisor tests`)
- [x] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

File: supervisor/dbus/network/connection.py [hunk 1]
```diff
@@ -104,31 +129,49 @@ def state(self) -> int:
...
+                for nameserver in ip4_data.get(DBUS_ATTR_NAMESERVER_DATA, [])
+            ],
+            [
+                ip_interface(f"{address[ATTR_ADDRESS]}/{address[ATTR_PREFIX]}")
+                for address in ip4_data.get(DBUS_ATTR_ADDRESS_DATA, [])
+            ],
+        )
+        self._ip6_config = IpConfiguration(
+            ip_address(ip6_data.get(DBUS_ATTR_GATEWAY))
+            if DBUS_ATTR_GATEWAY in ip4_data
```
Comment (ludeeus) [inline supervisor/dbus/network/connection.py:164]:
```suggestion
            if DBUS_ATTR_GATEWAY in ip6_data and ip6_data.get(DBUS_ATTR_GATEWAY)
```

File: supervisor/dbus/network/connection.py [hunk 2]
```diff
@@ -104,31 +129,49 @@ def state(self) -> int:
...
         self._ip4_config = IpConfiguration(
-            ip4_data.get(DBUS_ATTR_GATEWAY),
-            data[ATTR_IPV4].get(ATTR_METHOD),
-            ip4_data.get(DBUS_ATTR_NAMESERVERS),
-            AddressData(
-                ip4_data.get(DBUS_ATTR_ADDRESS_DATA)[0].get(ATTR_ADDRESS),
-                ip4_data.get(DBUS_ATTR_ADDRESS_DATA)[0].get(ATTR_PREFIX),
-            ),
+            ip_address(ip4_data[DBUS_ATTR_GATEWAY])
+            if DBUS_ATTR_GATEWAY in ip4_data
```
Comment (ludeeus) [inline supervisor/dbus/network/connection.py:150]:
```suggestion
            if DBUS_ATTR_GATEWAY in ip4_data and ip4_data.get(DBUS_ATTR_GATEWAY)
```

File: supervisor/dbus/network/connection.py [hunk 3]
```diff
@@ -78,21 +91,33 @@ def type(self) -> str:
 
     @property
     def ip4_config(self) -> IpConfiguration:
-        """Return a ip configuration object for the connection."""
+        """Return a ip6 configuration object for the connection."""
```
Comment (ludeeus) [inline supervisor/dbus/network/connection.py:94]:
```suggestion
        """Return a ip4 configuration object for the connection."""
```

**Review by ludeeus — APPROVED** [review]
Nice work! :tada:

File: supervisor/host/network.py [hunk 4]
```diff
@@ -44,3 +78,237 @@ async def update(self):
...
+    async def scan_wifi(self, interface: Interface) -> List[AccessPoint]:
+        """Scan on Interface for AccessPoint."""
+        inet = self.sys_dbus.network.interfaces.get(interface.name)
+
+        if inet.type != DeviceType.WIRELESS:
+            _LOGGER.error("Can only scan with wireless card - %s", interface.name)
+            raise HostNotSupportedError()
+
+        await inet.wireless.request_scan()
+        await asyncio.sleep(5)
```
Comment (ludeeus) [inline supervisor/host/network.py:141]: Should this be configurable?
In a chatty env, 5s might not be enough

**Review by ludeeus — CHANGES_REQUESTED** [review]

File: supervisor/api/network.py [hunk 5]
```diff
@@ -1,74 +1,158 @@
...
+    }
+
+
+def interface_struct(interface: Interface) -> dict:
     """Return a dict with information of a interface to be used in th API."""
     return {
         ATTR_INTERFACE: interface.name,
-        ATTR_IP_ADDRESS: f"{interface.ip_address}/{interface.prefix}",
-        ATTR_GATEWAY: interface.gateway,
-        ATTR_ID: interface.id,
```
[the remark this answers is not in the document]
Comment (pvizeli) [reply supervisor/api/network.py:48]: Our ID is the network interface. We do not try to cover nm, we have our own handling

File: supervisor/host/network.py (hunk 4 above)
Comment (pvizeli) [reply supervisor/host/network.py:141]: Future, we listen to the dbus signal for update properties, but the PR is enough big. He scan also constantly in background. We have 100 warranty to see almost everything after 1sec. 5 should be fine until we add the signal listener

File: supervisor/dbus/network/connection.py [hunk 6]
```diff
@@ -1,146 +1,116 @@
...
+    def ipv6(self) -> Optional[IpConfiguration]:
+        """Return a ip6 configuration object for the connection."""
+        return self._ipv6
+
+    async def connect(self) -> None:
+        """Get connection information."""
+        self.dbus = await DBus.connect(DBUS_NAME_NM, self.object_path)
+        self.properties = await self.dbus.get_properties(DBUS_NAME_CONNECTION_ACTIVE)
+
+        # IPv4
```
[the remark this answers is not in the document]
Comment (pvizeli) [reply supervisor/dbus/network/connection.py:80]: That will be the question for all ipvx stuff. For only 2, I prefer this more readable and less generic way

File: supervisor/dbus/network/configuration.py [hunk 7]
```diff
@@ -1,71 +1,75 @@
...
+
+    assigned_mac: Optional[str] = attr.ib()
+
+
+@attr.s(slots=True)
+class VlanProperties:
+    """Ethernet properties object for Network Manager."""
+
+    id: Optional[int] = attr.ib()
+    parent: Optional[str] = attr.ib()
```
[the remark this answers is not in the document]
Comment (pvizeli) [reply supervisor/dbus/network/configuration.py:68]: This is the DBUS interface for NM. We adapt just the logic of NM. But all VLAN interface are own interface. The parent is only for config. They should never show or handle as child of an interface.

**Review by ludeeus — APPROVED** [review]
Needs updated documentation and potentially a dev-blog post that covers the breaking changes before it can be promoted to the beta channel

**Pull request merged**